### DISCUSSION TITLE
What are the necessary conditions to add DP AUX device ?

### DISCUSSION THREAD

**Discussion by YiJaneWu** [fwupd/fwupd#6776] [Q&A]
I try to add Display Port AUX device, I would like to ask some question about dpaux device control.
I know in the quirk USB list, It have to populate the quirk with all the VID/PID.

I noticed that both the plugin of `synaptics-mst` and `kinetic-dp` use dpaux
These devices use custom GUID values like board-ID, Chip-ID, device-kind... and also add `fu_dpaux_device_get_dpcd_ieee_oui` function to check device. 
Is it necessary to put these information(board-ID, Chip-ID, device-kind...) in the device before setup dpaux device?

What are the necessary conditions in `plugin_backend_device_added` for device to start?
I just want to test a kind of product, When do I know that the product is ready to connect dpaux after monitor actually plugged in? 

I am new to developing dpaux. Thanks so much for help.

**Comment by hughsie** [comment]
All good questions! I'll explain what fwupd does first, and that might explain a few of those answers along the way.

* At startup, fwupd scans for all dpaux devices and probes them
* Some plugins say "I care about dpaux devices" using `fu_plugin_add_device_udev_subsystem(plugin, "drm_dp_aux_dev")`
* In `plugin_class->backend_device_added` plugins get passed a `FuDpauxDevice` they can query (calling `fu_device_locker_new()` runs `FuDpauxDevice->probe()` and `FuDpauxDevice->setup()`, and we can use that information to check if there's a match (e.g. using the OUI in the quirk file) then the plugin creates a plugin-specific device that subclasses `FuDpauxDevice` which implements the specifics on how to do the firmware update, e.g. `klass_device->write_firmware` and friends.
* The new custom device is _added_ to the daemon which means it gets exported to the user

Does any of that help?

**Reply by YiJaneWu** [reply]
Sure! I Learn a lot from it.
I wondering if I use my test plugin without writing quirk file, It does’t set any oui, board_id, device_id..  in my device, may I directly transfer data to the device being plugged in.

In your recommendation, what is the best way to write a quirk file items for dpaux plugin, and it can identify different products.

**Reply by hughsie** [reply]
> if I use my test plugin without writing quirk file

Sure -- you don't *need* to use `FuDpauxDevice` as a subclass. Can you share how you identify and update your device now? richard_at_hughsie_dot_com if you'd rather use private email.
